File: src/teaselib/core/devices/remote/UDPMessage.java
package teaselib.core.devices.remote;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

/**
 * @author Citizen-Cane@github.com
 *         <p>
 *         Message format:
 *         <p>
 *         <ul>
 *         <li>short text part size : the size of the text part (without this size info)
 *         <li>byte parameter count : number of parameters after the name
 *         <li>String command: the command name, null-terminated
 *         <li>List<String> parameters: all parameters, null-terminated
 *         <li>short binary size: The size of the binary part of the message (without this size info)
 *         <li>byte[size] binary data: the binary data of the message
 *         </ul>
 *         <p>
 */
public class UDPMessage {
    private static final String Encoding = "UTF-8";

    public final RemoteDeviceMessage message;

    static boolean isValid(byte[] data, int startIndex) {
        int offset = 0;
        int textSize = 256 * data[startIndex + offset++] + data[startIndex + offset++];
        int parameterCount = data[startIndex + offset++];
        int commandNameLength = strlen(data, startIndex + offset);
        offset += commandNameLength + 1;
        for (int i = 0; i < parameterCount; i++) {
            int parameterLength = strlen(data, startIndex + offset);
            offset += parameterLength + 1;
        }
        if (offset - 2 != textSize) {
            return false;
        }
        if (data.length >= offset + 2) {
            int binarySize = 256 * data[startIndex + offset++] + data[startIndex + offset++];
            if (data.length < offset + binarySize) {
                return false;
            }
            offset += binarySize;
        }
        return data.length == offset;
    }

    /**
     * Return string length in byte array
     * 
     * @param data
     * @param i
     * @return The number of bytes to the next 0
     */
    private static int strlen(byte[] data, int start) {
        int n = 0;
        for (int i = start; i < data.length; i++) {
            if (data[i] == 0)
                break;
            n++;
        }
        return n;
    }

    public UDPMessage(RemoteDeviceMessage message) {
        this.message = message;
    }

    public UDPMessage(String command, String... parameters) {
        this(command, Arrays.asList(parameters), new byte[0]);

    }

    public UDPMessage(String command, List<String> parameters, byte[] binary) {
        this.message = new RemoteDeviceMessage(command, parameters, binary);
    }

    public UDPMessage(byte[] data) throws IOException {
        ByteArrayInputStream input = new ByteArrayInputStream(data);
        DataInputStream dataInputStream = new DataInputStream(input);
        int textDataSize = dataInputStream.readShort();
        byte[] textData = new byte[textDataSize - 1];
        int parameterCount = dataInputStream.readByte();
        dataInputStream.readFully(textData);
        Scanner scanner = new Scanner(new ByteArrayInputStream(textData), Encoding);
        try {
            scanner.useDelimiter("\u0000");
            String name = scanner.next();
            List<String> parameters = new ArrayList<>(parameterCount);
            while (scanner.hasNext() && parameters.size() < parameterCount) {
                parameters.add(scanner.next());
            }
            int binarySize = dataInputStream.readShort();
            byte[] binary = new byte[binarySize];
            try {
                if (binarySize > 0) {
                    dataInputStream.readFully(binary);
                }
            } finally {
                dataInputStream.close();
            }
            message = new RemoteDeviceMessage(name, parameters, binary);
        } finally {
            scanner.close();
        }
    }

    public byte[] toByteArray() throws IOException {
        ByteArrayOutputStream data = new ByteArrayOutputStream();
        DataOutputStream output = new DataOutputStream(data);
        byte[] textData = getTextData();
        output.writeShort(textData.length);
        output.write(textData);
        output.writeShort(message.binary.length);
        if (message.binary.length > 0) {
            output.write(message.binary);
        }
        data.close();
        return data.toByteArray();
    }

    private byte[] getTextData() throws IOException {
        ByteArrayOutputStream textData = new ByteArrayOutputStream();
        DataOutputStream textDataOuptut = new DataOutputStream(textData);
        textDataOuptut.writeByte(message.parameters.size());
        textDataOuptut.write(message.command.getBytes());
        textDataOuptut.writeByte(0);
        for (String parameter : message.parameters) {
            textDataOuptut.write(parameter.getBytes());
            textDataOuptut.write(0);
        }
        return textData.toByteArray();
    }
}

File: firmware/particle/photon/TeaseLibUDPDevice/UDPMessage.cpp
#include "string.h"

#include "UDPMessage.h"


int UDPMessage::readShort(const char* data) {
    return 256 * data[0] + data[1];
}

void UDPMessage::writeShort(char* buffer, const int value) {
  buffer[0] = (value & 0xffff0000) >> 8;
  buffer[1] = (value & 0x0000ffff);
}

int UDPMessage::sizeOf(const char* buffer) {
  const int textPartSize = readShort(buffer);
  const int binarySize = readShort(&buffer[2 + textPartSize]);
  return 2 + textPartSize + 2 + binarySize;
}

int UDPMessage::sizeOf(const char* command, const char** parameters, const int parameterCount) {
  int size = strlen(command) + 1;
  for(int i = 0; i < parameterCount; i++) {
    size += strlen(parameters[i]) + 1;
  }
  return size;
}

int UDPMessage::sizeOf(const char* command, const char** parameters, const int parameterCount, const int binarySize) {
  return 2 + 1 + sizeOf(command, parameters, parameterCount) + 2 + binarySize;
}

bool UDPMessage::isValid(const char* data, const int size, int startIndex) {
  int offset = 0;
  int textSize = readShort(&data[startIndex + offset]);
  offset +=2;
  int parameterCount = data[startIndex + offset++];
  int commandNameLength = strlen(&data[startIndex + offset]);
  offset += commandNameLength + 1;
  for (int i = 0; i < parameterCount; i++) {
      int parameterLength = strlen(&data[startIndex + offset]);
      offset += parameterLength + 1;
  }
  if (offset - 2 != textSize) {
      return false;
  }
  if (size >= offset + 2) {
      int binarySize = readShort(&data[startIndex + offset]);
      offset +=2;
      if (size < offset + binarySize) {
          return false;
      }
      offset += binarySize;
  }
  return offset == size;
}

UDPMessage::UDPMessage(char* data, const int size)
: allocatedParameters(true) {
  const char* reader = data;
  const int textPartSize = readShort(reader);
  reader += 2;
  parameterCount = *reader++;
  command = reader;
  reader += strlen(reader) + 1;
  parameters = new const char*[parameterCount];
  for(int i = 0; i < parameterCount; i++) {
    parameters[i] = reader;
    reader += strlen(reader) + 1;
  }
  binarySize = readShort(reader);
  reader += 2;
  if (binarySize > 0) {
    binary = reader;
  } else {
    binary = NULL;
  }
}

UDPMessage::UDPMessage(const char* command, const char** parameters, const int parameterCount)
: command(command), parameters(parameters), parameterCount(parameterCount), binary(NULL), binarySize(0), allocatedParameters(false) {
}

UDPMessage::UDPMessage(const char* command, const char** parameters, const int parameterCount, const void* binary, const int binarySize)
: command(command), parameters(parameters), parameterCount(parameterCount), binary(binary), binarySize(binarySize), allocatedParameters(false)  {
}

int UDPMessage::toBuffer(char* buffer) {
  int index = 0;
  writeShort(buffer + index, sizeOf(command, parameters, parameterCount));
  index += 2;
  strcpy(buffer + index, command);
  index += strlen(command);
  buffer[index++] = 0x0;
  for(int i = 0; i < parameterCount; i++) {
    strcpy(buffer + index, parameters[i]);
    index += strlen(parameters[i]);
    buffer[index++] = 0x0;
    writeShort(buffer + index, binarySize);
    index += 2;
    if (binarySize > 0) {
      memcpy(buffer + index, binary, binarySize);
      index += binarySize;
    }
  }
  return index;
}

UDPMessage::~UDPMessage() {
  if (allocatedParameters) {
    delete parameters;
  }
}
